File: scripts/test_connection.py
#!/usr/bin/env python3
"""
LDAP Connection Test Script
Demonstrates how to connect, authenticate, and search an LDAP directory.

Requirements:
    pip install ldap3

Usage:
    python test_connection.py
"""

from ldap3 import Server, Connection, ALL, SUBTREE
import sys


# LDAP Configuration
LDAP_HOST = "localhost"
LDAP_PORT = 389
LDAP_USE_SSL = False
BASE_DN = "dc=example,dc=com"
ADMIN_DN = "cn=admin,dc=example,dc=com"
ADMIN_PASSWORD = "admin123"


def get_connection(user_dn=None, password=None):
    """Create and return an LDAP connection."""
    server = Server(
        f"ldap://{LDAP_HOST}:{LDAP_PORT}",
        get_info=ALL
    )
    
    if user_dn and password:
        conn = Connection(server, user=user_dn, password=password, auto_bind=True)
    else:
        conn = Connection(server, auto_bind=True)  # Anonymous bind
    
    return conn


def test_admin_connection():
    """Test admin authentication."""
    print("=" * 60)
    print("TEST 1: Admin Connection")
    print("=" * 60)
    
    try:
        conn = get_connection(ADMIN_DN, ADMIN_PASSWORD)
        print(f"✅ Successfully connected as admin")
        print(f"   Server: {conn.server.host}")
        print(f"   User: {ADMIN_DN}")
        conn.unbind()
        return True
    except Exception as e:
        print(f"❌ Failed to connect: {e}")
        return False


def test_user_authentication(username, password):
    """Test user authentication."""
    print("\n" + "=" * 60)
    print(f"TEST 2: User Authentication - {username}")
    print("=" * 60)
    
    user_dn = f"cn={username},ou=users,{BASE_DN}"
    
    try:
        conn = get_connection(user_dn, password)
        print(f"✅ Authentication successful for {username}")
        conn.unbind()
        return True
    except Exception as e:
        print(f"❌ Authentication failed: {e}")
        return False


def search_all_users():
    """Search and list all users."""
    print("\n" + "=" * 60)
    print("TEST 3: Search All Users")
    print("=" * 60)
    
    try:
        conn = get_connection(ADMIN_DN, ADMIN_PASSWORD)
        
        # Search for all users
        conn.search(
            search_base=f"ou=users,{BASE_DN}",
            search_filter="(objectClass=inetOrgPerson)",
            search_scope=SUBTREE,
            attributes=['cn', 'mail', 'uid', 'displayName', 'title']
        )
        
        print(f"✅ Found {len(conn.entries)} users:\n")
        
        for entry in conn.entries:
            print(f"   👤 {entry.displayName}")
            print(f"      DN: {entry.entry_dn}")
            print(f"      Email: {entry.mail}")
            print(f"      UID: {entry.uid}")
            if hasattr(entry, 'title') and entry.title:
                print(f"      Title: {entry.title}")
            print()
        
        conn.unbind()
        return True
    except Exception as e:
        print(f"❌ Search failed: {e}")
        return False


def search_by_filter(search_filter):
    """Search with a custom filter."""
    print("\n" + "=" * 60)
    print(f"TEST 4: Custom Filter Search")
    print(f"Filter: {search_filter}")
    print("=" * 60)
    
    try:
        conn = get_connection(ADMIN_DN, ADMIN_PASSWORD)
        
        conn.search(
            search_base=BASE_DN,
            search_filter=search_filter,
            search_scope=SUBTREE,
            attributes=['cn', 'mail', 'objectClass']
        )
        
        print(f"✅ Found {len(conn.entries)} entries:\n")
        
        for entry in conn.entries:
            print(f"   📁 {entry.entry_dn}")
        
        conn.unbind()
        return True
    except Exception as e:
        print(f"❌ Search failed: {e}")
        return False


def list_groups():
    """List all groups and their members."""
    print("\n" + "=" * 60)
    print("TEST 5: List Groups")
    print("=" * 60)
    
    try:
        conn = get_connection(ADMIN_DN, ADMIN_PASSWORD)
        
        conn.search(
            search_base=f"ou=groups,{BASE_DN}",
            search_filter="(objectClass=groupOfNames)",
            search_scope=SUBTREE,
            attributes=['cn', 'description', 'member']
        )
        
        print(f"✅ Found {len(conn.entries)} groups:\n")
        
        for entry in conn.entries:
            print(f"   👥 {entry.cn}")
            print(f"      Description: {entry.description}")
            members = entry.member if hasattr(entry, 'member') else []
            print(f"      Members: {len(members)}")
            print()
        
        conn.unbind()
        return True
    except Exception as e:
        print(f"❌ Failed to list groups: {e}")
        return False


def check_group_membership(username, group_name):
    """Check if a user belongs to a specific group."""
    print("\n" + "=" * 60)
    print(f"TEST 6: Check Group Membership")
    print(f"User: {username}, Group: {group_name}")
    print("=" * 60)
    
    try:
        conn = get_connection(ADMIN_DN, ADMIN_PASSWORD)
        
        user_dn = f"cn={username},ou=users,{BASE_DN}"
        group_dn = f"cn={group_name},ou=groups,{BASE_DN}"
        
        conn.search(
            search_base=group_dn,
            search_filter=f"(member={user_dn})",
            search_scope=SUBTREE
        )
        
        if conn.entries:
            print(f"✅ {username} IS a member of {group_name}")
        else:
            print(f"ℹ️  {username} is NOT a member of {group_name}")
        
        conn.unbind()
        return True
    except Exception as e:
        print(f"❌ Check failed: {e}")
        return False


def main():
    """Run all tests."""
    print("\n🔐 LDAP Connection Test Suite")
    print("=" * 60)
    print(f"Server: ldap://{LDAP_HOST}:{LDAP_PORT}")
    print(f"Base DN: {BASE_DN}")
    print("=" * 60)
    
    # Run tests
    test_admin_connection()
    test_user_authentication("john.doe", "password123")
    search_all_users()
    search_by_filter("(uid=jane*)")
    list_groups()
    check_group_membership("jane.smith", "developers")
    
    print("\n" + "=" * 60)
    print("🏁 All tests completed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: scripts/test_connection.js
#!/usr/bin/env node
/**
 * LDAP Connection Test Script (Node.js)
 * Demonstrates how to connect, authenticate, and search an LDAP directory.
 * 
 * Requirements:
 *     npm install ldapjs
 * 
 * Usage:
 *     node test_connection.js
 */

const ldap = require('ldapjs');

// LDAP Configuration
const config = {
    url: 'ldap://localhost:389',
    baseDN: 'dc=example,dc=com',
    adminDN: 'cn=admin,dc=example,dc=com',
    adminPassword: 'admin123'
};

/**
 * Create an LDAP client
 */
function createClient() {
    return ldap.createClient({
        url: config.url,
        timeout: 5000,
        connectTimeout: 10000
    });
}

/**
 * Bind (authenticate) to LDAP
 */
function bind(client, dn, password) {
    return new Promise((resolve, reject) => {
        client.bind(dn, password, (err) => {
            if (err) {
                reject(err);
            } else {
                resolve();
            }
        });
    });
}

/**
 * Search LDAP directory
 */
function search(client, baseDN, options) {
    return new Promise((resolve, reject) => {
        const entries = [];

        client.search(baseDN, options, (err, res) => {
            if (err) {
                reject(err);
                return;
            }

            res.on('searchEntry', (entry) => {
                entries.push(entry.object);
            });

            res.on('error', (err) => {
                reject(err);
            });

            res.on('end', () => {
                resolve(entries);
            });
        });
    });
}

/**
 * Test 1: Admin Connection
 */
async function testAdminConnection() {
    console.log('='.repeat(60));
    console.log('TEST 1: Admin Connection');
    console.log('='.repeat(60));

    const client = createClient();

    try {
        await bind(client, config.adminDN, config.adminPassword);
        console.log('✅ Successfully connected as admin');
        console.log(`   Server: ${config.url}`);
        console.log(`   User: ${config.adminDN}`);
        client.unbind();
        return true;
    } catch (err) {
        console.log(`❌ Failed to connect: ${err.message}`);
        client.destroy();
        return false;
    }
}

/**
 * Test 2: User Authentication
 */
async function testUserAuthentication(username, password) {
    console.log('\n' + '='.repeat(60));
    console.log(`TEST 2: User Authentication - ${username}`);
    console.log('='.repeat(60));

    const client = createClient();
    const userDN = `cn=${username},ou=users,${config.baseDN}`;

    try {
        await bind(client, userDN, password);
        console.log(`✅ Authentication successful for ${username}`);
        client.unbind();
        return true;
    } catch (err) {
        console.log(`❌ Authentication failed: ${err.message}`);
        client.destroy();
        return false;
    }
}

/**
 * Test 3: Search All Users
 */
async function searchAllUsers() {
    console.log('\n' + '='.repeat(60));
    console.log('TEST 3: Search All Users');
    console.log('='.repeat(60));

    const client = createClient();

    try {
        await bind(client, config.adminDN, config.adminPassword);

        const searchOptions = {
            filter: '(objectClass=inetOrgPerson)',
            scope: 'sub',
            attributes: ['cn', 'mail', 'uid', 'displayName', 'title']
        };

        const users = await search(client, `ou=users,${config.baseDN}`, searchOptions);

        console.log(`✅ Found ${users.length} users:\n`);

        users.forEach((user) => {
            console.log(`   👤 ${user.displayName || user.cn}`);
            console.log(`      DN: ${user.dn}`);
            console.log(`      Email: ${user.mail}`);
            console.log(`      UID: ${user.uid}`);
            if (user.title) {
                console.log(`      Title: ${user.title}`);
            }
            console.log();
        });

        client.unbind();
        return true;
    } catch (err) {
        console.log(`❌ Search failed: ${err.message}`);
        client.destroy();
        return false;
    }
}

/**
 * Test 4: Custom Filter Search
 */
async function searchByFilter(filter) {
    console.log('\n' + '='.repeat(60));
    console.log('TEST 4: Custom Filter Search');
    console.log(`Filter: ${filter}`);
    console.log('='.repeat(60));

    const client = createClient();

    try {
        await bind(client, config.adminDN, config.adminPassword);

        const searchOptions = {
            filter: filter,
            scope: 'sub',
            attributes: ['cn', 'mail', 'objectClass']
        };

        const entries = await search(client, config.baseDN, searchOptions);

        console.log(`✅ Found ${entries.length} entries:\n`);

        entries.forEach((entry) => {
            console.log(`   📁 ${entry.dn}`);
        });

        client.unbind();
        return true;
    } catch (err) {
        console.log(`❌ Search failed: ${err.message}`);
        client.destroy();
        return false;
    }
}

/**
 * Test 5: List Groups
 */
async function listGroups() {
    console.log('\n' + '='.repeat(60));
    console.log('TEST 5: List Groups');
    console.log('='.repeat(60));

    const client = createClient();

    try {
        await bind(client, config.adminDN, config.adminPassword);

        const searchOptions = {
            filter: '(objectClass=groupOfNames)',
            scope: 'sub',
            attributes: ['cn', 'description', 'member']
        };

        const groups = await search(client, `ou=groups,${config.baseDN}`, searchOptions);

        console.log(`✅ Found ${groups.length} groups:\n`);

        groups.forEach((group) => {
            console.log(`   👥 ${group.cn}`);
            console.log(`      Description: ${group.description}`);
            const members = Array.isArray(group.member) ? group.member : [group.member];
            console.log(`      Members: ${members.length}`);
            console.log();
        });

        client.unbind();
        return true;
    } catch (err) {
        console.log(`❌ Failed to list groups: ${err.message}`);
        client.destroy();
        return false;
    }
}

/**
 * Main function to run all tests
 */
async function main() {
    console.log('\n🔐 LDAP Connection Test Suite (Node.js)');
    console.log('='.repeat(60));
    console.log(`Server: ${config.url}`);
    console.log(`Base DN: ${config.baseDN}`);
    console.log('='.repeat(60));

    await testAdminConnection();
    await testUserAuthentication('john.doe', 'password123');
    await searchAllUsers();
    await searchByFilter('(uid=jane*)');
    await listGroups();

    console.log('\n' + '='.repeat(60));
    console.log('🏁 All tests completed!');
    console.log('='.repeat(60));
}

main().catch(console.error);
